### ISSUE TITLE
Unable to open project with cloned group nodes (segfault)

### ISSUE THREAD

**Issue by josegomezr** [NatronGitHub/Natron#568]
Hi there,

First of all, congrats for this awesome product :tada:! I enjoyed using it for the first time, watched a couple of Youtube Videos and got straight ahead with some simple results. 

I ran into this problem while doing some practicing on my own, I've provided screenshots to replicate it too.

### Problem

When using "Clone Nodes" feature, Natron will not be able to open the saved project.

**Expected behavior:** It should open the project in the last saved state

**Actual behavior:** Crashes when opening a project with cloned nodes

 * If your problem can be reproduced using a Natron project, please include a link to the project on a file sharing service, or attach the project as a zip file to this issue, if possible.
 
[broken-project.tar.gz](https://github.com/NatronGitHub/Natron/files/5937771/broken-project.tar.gz)

 * If your problem is a crash in an official release/snapshot, please include verbose output from the application
 from a terminal if possible.  If you also submitted a crash report, indicate the CrashID if possible.

**Unfortunately** it just segfaults:

![image](https://user-images.githubusercontent.com/5963207/107130713-ab381e00-68d0-11eb-8220-4d21fccee217.png)

### Steps to Reproduce

1. Open a new project
2. Pull two rectangles + a merge node as such:
  ![image](https://user-images.githubusercontent.com/5963207/107130615-8b542a80-68cf-11eb-8c04-ebebafd499ba.png)
3. Save, Close & Re-open: works as expected. No problem so far.
4. Group those rectangles + Merge into one node:
  ![image](https://user-images.githubusercontent.com/5963207/107130639-d5d5a700-68cf-11eb-9514-e91a68176986.png)
5. Save, Close & Re-open: works as expected. No problem so far.
6. Clone Group1 and add a Transform & Merge Node (so we can shift it in another direction, just to see the 4 boxes). ex:
  ![image](https://user-images.githubusercontent.com/5963207/107130665-2c42e580-68d0-11eb-997d-855af4a5ddb8.png)
7. Save, Close & Re-open
8. Crashed.

You may submit a link to any screenshots/videos that can be used to understand how to reproduce the issue.

### Versions

* **Natron version/commit** (they can be retrieved from the About Window): Natron version 2.3.15 at commit 513a2491008c152417ae913f8d1ab83e432aa7e3 on branch tags/v2.3.15

* **OS version:** openSUSE Leap 15.2

**Comment by devernay** [comment]
Good bug report, thanks! Will look into it when I or Ole-André have time!

**Comment by rodlie** [comment]
``Engine/Knob.cpp:3513: virtual bool Natron::KnobHelper::slaveToInternal(int, const KnobIPtr&, int, Natron::ValueChangedReasonEnum, bool): Assertion `other.get() != this' failed.``
```
bool
KnobHelper::slaveToInternal(int dimension,
                            const KnobIPtr & other,
                            int otherDimension,
                            ValueChangedReasonEnum reason,
                            bool ignoreMasterPersistence)
{
    assert(other.get() != this);
```

**Comment by josegomezr** [comment]
does that means that assertion is superfluous? or the problem is else where and happens to be caught there? 🤔

**Comment by rodlie** [comment]
If I remove the assert we end up in ``readWriteLock()->lockForRead();``
```
1   QReadWriteLock::lockForRead()                       0x7ffff6a10101 
2   QReadLocker::relock           qreadwritelock.h 111  0x555556eb582f 
3   QReadLocker::QReadLocker      qreadwritelock.h 130  0x555556eb58c2 
4   Natron::KnobHelper::getMaster Knob.cpp         3583 0x5555575baa55 
5   Natron::KnobHelper::getCurve  Knob.cpp         1580 0x5555575b0c4a 
6   Natron::KnobHelper::getCurve  Knob.cpp         1582 0x5555575b0cac 
7   Natron::KnobHelper::getCurve  Knob.cpp         1582 0x5555575b0cac 
8   Natron::KnobHelper::getCurve  Knob.cpp         1582 0x5555575b0cac 
9   Natron::KnobHelper::getCurve  Knob.cpp         1582 0x5555575b0cac 
10  Natron::KnobHelper::getCurve  Knob.cpp         1582 0x5555575b0cac 
11  Natron::KnobHelper::getCurve  Knob.cpp         1582 0x5555575b0cac 
12  Natron::KnobHelper::getCurve  Knob.cpp         1582 0x5555575b0cac 
13  Natron::KnobHelper::getCurve  Knob.cpp         1582 0x5555575b0cac 
14  Natron::KnobHelper::getCurve  Knob.cpp         1582 0x5555575b0cac 
15  Natron::KnobHelper::getCurve  Knob.cpp         1582 0x5555575b0cac 
16  Natron::KnobHelper::getCurve  Knob.cpp         1582 0x5555575b0cac 
17  Natron::KnobHelper::getCurve  Knob.cpp         1582 0x5555575b0cac 
18  Natron::KnobHelper::getCurve  Knob.cpp         1582 0x5555575b0cac 
19  Natron::KnobHelper::getCurve  Knob.cpp         1582 0x5555575b0cac 
20  Natron::KnobHelper::getCurve  Knob.cpp         1582 0x5555575b0cac 
... <More>                                                             
```
```
Loading project: /home/olear/tmp/tuto-1.ntp
Link slave/master for  recenter  failed to restore the following linkage:  Rectangle1_2 
Link slave/master for  recenter  failed to restore the following linkage:  Rectangle1 
Link slave/master for  screenAlpha_separator  failed to restore the following linkage:  Merge1 
Link slave/master for  OutputChannelsA_separator  failed to restore the following linkage:  Merge1 
```

**Comment by josegomezr** [comment]
looks like a more complicated issue 😅

Maybe is related on how the XML is built? because the clone node works perfectly in the project, the problem arises when reading the project.

I wish I had Qt expertise to help debugging.

**Comment by rodlie** [comment]
Cloning any group (regardless of what they contain) seems to trigger this.

**Comment by josegomezr** [comment]
bump! Can I help with this bug? 😅

**Comment by rodlie** [comment]
The issue can easily be replicated, so we don't need any additional help on that. I might have some time tomorrow.

**Comment by rodlie** [comment]
I have a fix, will submit tonight or tomorrow.

**Comment by devernay** [comment]
@rodlie can you make a PR with the fix?

**Comment by devernay** [comment]
OK I started the work in https://github.com/NatronGitHub/Natron/tree/fix-issue-568

It doesn't crash anymore, but group clones cannot be saved/loaded: links only work within the same group for now.

**Comment by devernay** [comment]
I got a solution that works for this situation (see https://github.com/NatronGitHub/Natron/pull/594), but links/clones are rather fragile. Here's the situation now:
- saving and loading a project with links/clones should always work (that's the nicest part: you will never lose your work)
- some operations may fail to undo/redo, and in this case links may be broken, but Natron shouldn't crash
- some operations may fail to restore links, for example expanding one of the two groups in your situation, or cloning clones...

**Comment by devernay** [comment]
Working on a slightly more backward-compatible solution

**Comment by devernay** [comment]
fixed by #598
